File: artifacts/zopfli/decompose/src/cache.c
void ZopfliInitCache(size_t blocksize, ZopfliLongestMatchCache *lmc)
{
  size_t i;
  lmc->length = (unsigned short *) malloc((sizeof(unsigned short)) * blocksize);
  lmc->dist = (unsigned short *) malloc((sizeof(unsigned short)) * blocksize);
  lmc->sublen = (unsigned char *) malloc((8 * 3) * blocksize);
  if (lmc->sublen == 0)
  {
    fprintf(stderr, "Error: Out of memory. Tried allocating %lu bytes of memory.\n", (((unsigned long) 8) * 3) * blocksize);
    exit(1);
  }
  for (i = 0; i < blocksize; i += 1)
    lmc->length[i] = 1;

  for (i = 0; i < blocksize; i += 1)
    lmc->dist[i] = 0;

  for (i = 0; i < ((8 * blocksize) * 3); i += 1)
    lmc->sublen[i] = 0;

}


/*** DEPENDENCIES:
typedef struct ZopfliLongestMatchCache
{
  unsigned short *length;
  unsigned short *dist;
  unsigned char *sublen;
} ZopfliLongestMatchCache
----------------------------
***/


void ZopfliCleanCache(ZopfliLongestMatchCache *lmc)
{
  free(lmc->length);
  free(lmc->dist);
  free(lmc->sublen);
}


/*** DEPENDENCIES:
typedef struct ZopfliLongestMatchCache
{
  unsigned short *length;
  unsigned short *dist;
  unsigned char *sublen;
} ZopfliLongestMatchCache
----------------------------
***/


unsigned ZopfliMaxCachedSublen(const ZopfliLongestMatchCache *lmc, size_t pos, size_t length)
{
  unsigned char *cache;
  unsigned int cache_idx = 0;
  cache_idx = &lmc->sublen[(8 * pos) * 3];
  (void) length;
  if ((cache[1 + cache_idx] == 0) && (cache[2 + cache_idx] == 0))
  {
    return 0;
  }
  return cache[((8 - 1) * 3) + cache_idx] + 3;
}


/*** DEPENDENCIES:
typedef struct ZopfliLongestMatchCache
{
  unsigned short *length;
  unsigned short *dist;
  unsigned char *sublen;
} ZopfliLongestMatchCache
----------------------------
***/


void ZopfliSublenToCache(const unsigned short *sublen, size_t pos, size_t length, ZopfliLongestMatchCache *lmc)
{
  size_t i;
  size_t j = 0;
  unsigned bestlength = 0;
  unsigned char *cache;
  unsigned int cache_idx = 0;
  cache_idx = &lmc->sublen[(8 * pos) * 3];
  if (length < 3)
  {
    return;
  }
  for (i = 3; i <= length; i += 1)
  {
    if ((i == length) || (sublen[i] != sublen[i + 1]))
    {
      cache[(j * 3) + cache_idx] = i - 3;
      cache[((j * 3) + 1) + cache_idx] = sublen[i] % 256;
      cache[((j * 3) + 2) + cache_idx] = (sublen[i] >> 8) % 256;
      bestlength = i;
      j += 1;
      if (j >= 8)
      {
        break;
      }
    }
  }

  if (j < 8)
  {
    assert(bestlength == length);
    cache[((8 - 1) * 3) + cache_idx] = bestlength - 3;
  }
  else
  {
    assert(bestlength <= length);
  }
  assert(bestlength == ZopfliMaxCachedSublen(lmc, pos, length));
}


/*** DEPENDENCIES:
typedef struct ZopfliLongestMatchCache
{
  unsigned short *length;
  unsigned short *dist;
  unsigned char *sublen;
} ZopfliLongestMatchCache
----------------------------
unsigned ZopfliMaxCachedSublen(const ZopfliLongestMatchCache *lmc, size_t pos, size_t length)
{
  unsigned char *cache;
  unsigned int cache_idx = 0;
  cache_idx = &lmc->sublen[(8 * pos) * 3];
  (void) length;
  if ((cache[1 + cache_idx] == 0) && (cache[2 + cache_idx] == 0))
  {
    return 0;
  }
  return cache[((8 - 1) * 3) + cache_idx] + 3;
}


----------------------------
***/


void ZopfliCacheToSublen(const ZopfliLongestMatchCache *lmc, size_t pos, size_t length, unsigned short *sublen)
{
  size_t i;
  size_t j;
  unsigned maxlength = ZopfliMaxCachedSublen(lmc, pos, length);
  unsigned prevlength = 0;
  unsigned char *cache;
  unsigned int cache_idx = 0;
  if (length < 3)
  {
    return;
  }
  cache_idx = &lmc->sublen[(8 * pos) * 3];
  for (j = 0; j < 8; j += 1)
  {
    unsigned length = cache[(j * 3) + cache_idx] + 3;
    unsigned dist = cache[((j * 3) + 1) + cache_idx] + (256 * cache[((j * 3) + 2) + cache_idx]);
    for (i = prevlength; i <= length; i += 1)
    {
      sublen[i] = dist;
    }

    if (length == maxlength)
    {
      break;
    }
    prevlength = length + 1;
  }

}


/*** DEPENDENCIES:
typedef struct ZopfliLongestMatchCache
{
  unsigned short *length;
  unsigned short *dist;
  unsigned char *sublen;
} ZopfliLongestMatchCache
----------------------------
unsigned ZopfliMaxCachedSublen(const ZopfliLongestMatchCache *lmc, size_t pos, size_t length)
{
  unsigned char *cache;
  unsigned int cache_idx = 0;
  cache_idx = &lmc->sublen[(8 * pos) * 3];
  (void) length;
  if ((cache[1 + cache_idx] == 0) && (cache[2 + cache_idx] == 0))
  {
    return 0;
  }
  return cache[((8 - 1) * 3) + cache_idx] + 3;
}


----------------------------
***/

File: artifacts/zopfli/translate/src/cache.rs
use crate::*;

pub fn ZopfliInitCache(blocksize: usize, lmc: &mut ZopfliLongestMatchCache) {
    // Allocate vectors with capacity `blocksize` for length and dist
    lmc.length = Some(vec![1; blocksize]);
    lmc.dist = Some(vec![0; blocksize]);
    
    // Calculate sublen size (8 * 3 * blocksize)
    let sublen_size = 8 * 3 * blocksize;
    lmc.sublen = Some(vec![0; sublen_size]);
    
    // In Rust, Vec allocation will panic if out of memory, so no need for explicit error checking
    // The initialization values are set directly in the vec! macro calls
}
pub fn ZopfliCleanCache(lmc: &mut ZopfliLongestMatchCache) {
    // In Rust, dropping the Vecs (by setting them to None) will automatically free their memory
    lmc.length = None;
    lmc.dist = None;
    lmc.sublen = None;
}
pub fn ZopfliMaxCachedSublen(lmc: &ZopfliLongestMatchCache, pos: usize, length: usize) -> u8 {
    // The `length` parameter is unused, so we silence the warning
    let _ = length;
    
    // Safely access the sublen vector
    let cache = match &lmc.sublen {
        Some(c) => c,
        None => return 0,
    };

    // Calculate the cache index
    let cache_idx = (8 * pos) * 3;
    
    // Check if we can safely access the required indices
    if cache_idx + 2 >= cache.len() {
        return 0;
    }

    // Check the conditions from the C code
    if cache[cache_idx + 1] == 0 && cache[cache_idx + 2] == 0 {
        return 0;
    }

    // Calculate the final index and return the result
    let final_idx = ((8 - 1) * 3) + cache_idx;
    if final_idx >= cache.len() {
        return 0;
    }
    
    cache[final_idx] + 3
}
pub fn ZopfliSublenToCache(sublen: &[u16], pos: usize, length: usize, lmc: &mut ZopfliLongestMatchCache) {
    let mut j = 0;
    let mut bestlength = 0;
    
    // Early return if length is less than 3
    if length < 3 {
        return;
    }

    // Get mutable access to the cache
    let cache = match &mut lmc.sublen {
        Some(c) => c,
        None => return,
    };

    let cache_idx = (8 * pos) * 3;

    for i in 3..=length {
        if i == length || sublen[i] != sublen[i + 1] {
            // Check bounds before writing
            if cache_idx + (j * 3) + 2 < cache.len() {
                cache[cache_idx + (j * 3)] = (i - 3) as u8;
                cache[cache_idx + (j * 3) + 1] = (sublen[i] % 256) as u8;
                cache[cache_idx + (j * 3) + 2] = ((sublen[i] >> 8) % 256) as u8;
            }
            bestlength = i;
            j += 1;
            if j >= 8 {
                break;
            }
        }
    }

    if j < 8 {
        assert!(bestlength == length);
        // Check bounds before writing
        if cache_idx + ((8 - 1) * 3) < cache.len() {
            cache[cache_idx + ((8 - 1) * 3)] = (bestlength - 3) as u8;
        }
    } else {
        assert!(bestlength <= length);
    }
    assert!(bestlength == ZopfliMaxCachedSublen(lmc, pos, length) as usize);
}
pub fn ZopfliCacheToSublen(lmc: &ZopfliLongestMatchCache, pos: usize, length: usize, sublen: &mut [u16]) {
    if length < 3 {
        return;
    }

    let maxlength = ZopfliMaxCachedSublen(lmc, pos, length);
    let mut prevlength = 0;

    let cache = match &lmc.sublen {
        Some(c) => c,
        None => return,
    };

    let cache_idx = (8 * pos) * 3;
    if cache_idx >= cache.len() {
        return;
    }

    for j in 0..8 {
        let j_offset = (j * 3) + cache_idx;
        if j_offset + 2 >= cache.len() {
            break;
        }

        let length_val = cache[j_offset] + 3;
        let dist = cache[j_offset + 1] as u16 + (256 * cache[j_offset + 2] as u16);

        for i in prevlength..=length_val {
            if usize::from(i) < sublen.len() {
                sublen[i as usize] = dist;
            }
        }

        if length_val == maxlength {
            break;
        }
        prevlength = length_val + 1;
    }
}
